README.md
# Haskell Practice Directories (99hs) #

Here is a folder where I will write solutions to H-99: the ninety-nine Haskell
problems found on
[https://wiki.haskell.org/H-99:_Ninety-Nine_Haskell_Problems](wiki.haskell.org).


ChangeLog.md
# Changelog for 99hs

## Unreleased changes


test/Spec.hs
import qualified Q1t10
import qualified Q11t20
import qualified Q21t28
import qualified Q31t41

main :: IO ()
main = do
    let failQ1t10 = head [n | n <- [1..], not (Q1t10.tests n)]
    if failQ1t10 <= 10 then 
        error $ "Test failed at Q" ++ show failQ1t10
            else putStrLn "Q1-10 Passed!"
    let failQ11t20 = head [n | n <- [11..], not (Q11t20.tests n)]
    if failQ11t20 <= 20 then
        error $ "Test failed at Q" ++ show failQ11t20
            else putStrLn "Q11-20 Passed!"
    let failQ21t28 = head [n | n <- [21..], not (Q21t28.tests n)]
    if failQ21t28 <= 28 then
        error $ "Test failed at Q" ++ show failQ21t28
            else putStrLn "Q21-28 Passed!"
    let failQ31t41 = head [n | n <- [31..], not (Q31t41.tests n)]
    if failQ31t41 <= 41 then
        error $ "Test failed at Q" ++ show failQ31t41
            else putStrLn "Q31-41 Passed!"
    putStrLn "Tests complete!"


src/Q1t10.hs
module Q1t10 where

{- |The unit tests for Questions 1-10 -}
tests :: Int -> Bool
tests 1 = myLast [1,2,3,4] == Just 4 && myLast ['x','y','z'] == Just 'z'
tests 2 = penultimate [1,2,3,4] == Just 3 && penultimate ['a'..'z'] == Just 'y'
tests 3 = kth [1,2,3] 2 == Just 2 && kth "haskell" 5 == Just 'e'
tests 4 = myLength [123,456,789] == 3 && myLength "Hello, world!" == 13
tests 5 = p && myReverse [1,2,3,4] == [4,3,2,1] where
    p = myReverse "A man, a plan, a canal. panama!" == "!amanap .lanac a ,nalp a ,nam A"
tests 6 = not (isPalindrome [1,2,3]) && isPalindrome "madamimadam" && 
    isPalindrome [1,2,4,8,26,8,4,2,1]
tests 7 = (flatten (Elem 5) == [5]) &&
    (flatten (List [Elem 1, List [Elem 2, List [Elem 3, Elem 4], Elem 5]]) ==
        [1,2,3,4,5]) && null (flatten (List []))
tests 8 = compress "aaaabccaadeeee" == "abcade"
tests 9 = pack "aaaabccaadeeee" == ["aaaa","b","cc","aa","d","eeee"]
tests 10 = encode "aaaabccaadeeee" ==
    [(4,Just 'a'),(1,Just 'b'),(2, Just 'c'),(2, Just 'a'),(1,Just 'd'),
        (4,Just 'e')]
tests _ = False -- If we reach this point then we are good.

-- Problem 1
-- Find the last element of a list
myLast :: [a] -> Maybe a
myLast [] = Nothing
myLast [x] = Just x
myLast (_:xs) = myLast xs

-- Problem 2
-- Find the penultimate element of a list
penultimate :: [a] -> Maybe a
penultimate [] = Nothing
penultimate [x] = Nothing
penultimate [x,_] = Just x
penultimate (x:xs) = penultimate xs

-- Problem 3
-- Find the k'th element of a list, where the first element has index 1
kth :: [a] -> Int -> Maybe a
kth _ k
    | k <= 0 = Nothing
kth [] _ = Nothing
kth x 1 = Just (head x)
kth x k = kth (tail x) (k-1)

-- Problem 4
-- Find the number of elements of a list
myLength :: [a] -> Int
myLength = foldr (\x -> (+) 1) 0 

-- Problem 5
-- Reverse a list
myReverse :: [a] -> [a]
myReverse = foldl (\y x0 -> x0:y) []
-- hlint notes we can replace (\y x0 -> x0:y) with flip (:)
-- If we do this, we match replace in the Prelude

-- Problem 6
-- Identify whether a list is a palindrome
isPalindrome :: (Eq a) => [a] -> Bool
isPalindrome x = x == myReverse x
-- Here's another weird way to do it (found on the solutions)
-- See https://stackoverflow.com/q/69936614/9302269
isPalindromeMonadic :: (Eq a) => [a] -> Bool
isPalindromeMonadic = reverse >>= (==)

-- Problem 7
-- Flatten a list of lists
data NestedList a = Elem a | List [NestedList a] 
-- Let's make things print nicely
instance Show a => Show (NestedList a) where
    show x = "(As NestedList) " ++ prep x where 
        prep (Elem x') = show x'
        prep (List xs) = "[" ++ head strs ++ concatMap (',':) (tail strs) ++ "]"
            where strs = map prep xs

makeNestedList :: [[a]] -> NestedList a
makeNestedList = List . map (List . map Elem)

flatten :: NestedList a -> [a]
flatten (Elem x) = [x]
flatten (List x) = foldr f [] x where
    f x' y' = flatten x' ++ y'

-- Problem 8
-- Eliminate consecutive duplicates of list elements
compress :: (Eq a) => [a] -> [a]
compress (x1:xs@(x2:_))
    | x1 == x2 = compress xs
    | otherwise = x1:compress xs
compress x = x
-- This works because haskell pattern matches from top of file downward.

-- Problem 9
-- Pack consecutive duplicates of list elements into sublists.
pack :: (Eq a) => [a] -> [[a]]
pack x = reverse $ foldl f [] x where
    f cks@(ck@(ck1:_):rest) x
        | x == ck1 = (x:ck):rest
        | otherwise = [x]:cks
    f _ x = [[x]] -- We only reach this case if _ is [] or [[]]
-- Is there a way to do this without using reverse?
-- Is there a way to write this so that it works for infinite lists?
-- One could use foldr, but then would pack work for infinite lists?
-- Here is something interesting from 
-- https://elbauldelprogramador.com/org-posts/foldr-infinite-list-haskell.html
-- """
--      A common misunderstanding with `foldr` and `foldl`, which I also had,
--      is that one may think the former start from the end of the list, and
--      the latter from the beginning. What really happens is that `foldl` is
--      left associative, and `foldr` right associative, and *both of them
--      start from the left most side* of the list.
-- """

-- Consider the follow solution which works for infinite lists:
pack' (x:xs) = let (first,rest) = span (==x) xs
    in (x:first) : pack' rest
pack' [] = []
-- The definition of span in Prelude is
span' :: (a -> Bool) -> [a] -> ([a],[a])
span' _ xs@[] = (xs,xs)
span' p xs@(x:xs')
    | p x = let (ys,zs) = span p xs' in (x:ys,zs)
    | otherwise = ([],xs)
-- Therefore there is something special about let and in that make the
-- infinite case work.
pack'' :: (Eq a) => [a] -> [[a]]
pack'' x = myReverse $ foldl f [] x where
    f cks@(ck@(ck1:_):rest) x
        | x == ck1 = (x:ck):rest
        | otherwise = [x]:cks
    f _ x = [[x]]
       
-- Problem 10
-- Using the result of Problem 9, implement run-length data compression.
protoEncode :: [[a]] -> [(Int, Maybe a)]
protoEncode = map f where
    f xs@(x:_) = (length xs,Just x)
    f [] = (0,Nothing)

encode :: (Eq a) => [a] -> [(Int,Maybe a)]
encode = protoEncode . pack


src/Q11t20.hs
module Q11t20 where
import Data.List (group)

tests :: Int -> Bool
tests 11 = encodeModified "aaaabccaadeeee" == 
    [Multiple 4 'a',Single 'b',Multiple 2 'c', Multiple 2'a',Single 'd',
        Multiple 4 'e']
tests 12 = decodeModified (encodeModified "aaaabccaadeeee") == "aaaabccaadeeee"
tests 13 = True -- This question was not implemented
tests 14 = dupli [1,2,3] == [1,1,2,2,3,3]
tests 15 = repli "abc" 3 == "aaabbbccc"
tests 16 = dropEvery "abcdefghik" 3 == "abdeghk"
tests 17 = split "abcdefghik" 3 == ("abc","defghik")
tests 18 = slice "abcdefghik" 3 7 == "cdefg"
tests 19 = myRotate "abcdefgh" 3 == "defghabc" &&
    myRotate "abcdefgh" (-2) == "ghabcdef"
tests 20 = removeAt 2 "abcd" == ('b',"acd")
tests _ = False -- We made it!

-- Problem 11
-- Implement run-length encoding, but if encoutered a single symbol, input it
-- directly into the encoded string rather than as a singleton set.
data Poly a = Single a | Multiple Int a deriving Eq

instance Show a => Show (Poly a) where
    show (Single x) = show x
    show (Multiple n x) = '(':show n++" "++show x++")"

embedList :: [a] -> [Poly a]
embedList = map Single

encodeModified :: (Eq a) =>  [a] -> [Poly a]
encodeModified = foldr (+!) []

(+!) :: (Eq a) => a -> [Poly a] -> [Poly a]
(+!) x [] = [Single x]
(+!) x ys@((Single y):ys')
    | x == y = Multiple 2 x : ys'
    | otherwise = Single x : ys
(+!) x ys@((Multiple n y):ys')
    | x == y = Multiple (n+1) x : ys' -- I think this line is why it hands for
                                      -- infinite lists, because it is changing
                                      -- the tail of ys, when we call foldr,
                                      -- GHC has to evaluate +! for the tail,
                                      -- which has to evaluate +! on the tail of
                                      -- the tail, and so on.
    | otherwise = Single x : ys
-- The above fails for infinite lists. I think this is because the pattern
-- matching forces GHC to evaluate the fold over the entire list.
encodeModifiedInf :: (Eq a) => [a] -> [Poly a]
encodeModifiedInf xs = 
    [y | x <- group xs, 
        let y = if 
            length x == 1 then Single (head x) 
                else Multiple (length x) (head x)]
-- DESCRIPTION OF LIST COMPREHENSION ABOVE "The elements y in the list are
-- given by taking the elements x of the list group xs where y is the
-- expression Single (head x) or Multiple (length x) (head x)"

-- Note this works for infinite lists for two reasons:
--  1) the constructor Multiple (length x) (head x) is called only once, rather
--     than modified n times as we zip through xs, and this is possible because
--     of (2)
--  2) `group` is lazy (because lists are lazily constructed), so the result of
--     encodeModifiedInf is not reduced until it is needed.

-- Patching encodeModified to work with infinite lists would amount to
-- rewriting `group`, albeit in a simpler notation (since `group` is built
-- from `groupExp`).

-- Problem 12
-- Decode the run-length encoding implemented above.
decodeModified :: [Poly a] -> [a]
decodeModified = foldr f [] where
    f :: Poly a -> [a] -> [a]
    f (Single x) = (:) x
    f (Multiple n x) = (++) [x | _ <- [1..n] ]

-- Problem 13
-- Do Problem 11 without creating the sublists (see Q1-10).
-- Since I did this already, this is moot.

-- Problem 14
-- Duplicate elements of a list.
dupli :: [a] -> [a]
dupli [] = []
dupli (x:xs) = x:x:dupli xs

-- Problem 15
-- Replicate the elements of a list n times, where n=1 is the identity.
repli :: [a] -> Int -> [a]
repli [] _ = []
repli (x:xs) n = [x | _ <- [1..n]] ++ repli xs n

-- Problem 16
-- Drop every n'th element from a list.
dropEvery :: [a] -> Int -> [a]
dropEvery xs = f xs 1 where
    f :: [a] -> Int -> Int -> [a]
    f [] _ _ = []
    f (x:xs) m n 
        | mod m n == 0 = f xs 1 n
        | otherwise = x:f xs (m+1) n

-- Problem 17
-- Split an list into two parts at index n

split :: [a] -> Int -> ([a],[a])
split ys n = (reverse left_side, right_side) where
    (left_side, right_side) = f n ([],ys) where
        -- f acts like a deck of cards, drawing from the right (top to bottom)
        -- and placing to the left (bottom to top).
        f :: Int -> ([a],[a]) -> ([a],[a])
        f 0 (xs,ys) = (xs,ys)
        f n (xs,y:ys) = f (n-1) (y:xs,ys) 
        f _ (xs,[]) = (xs,[])

-- Problem 18
-- Extract the slice from i to k in a list
slice :: [a] -> Int -> Int -> [a]
slice xs i k = fst (split (snd (split xs (i-1))) (k-i+1))

-- Problem 19
-- Rotate a list n places to the left
myRotate :: [a] -> Int -> [a]
myRotate xs n = let (l,r) = split xs m in r ++ l where
    m = mod n (length' xs) where
        length' = foldr (\_ y -> y+1) 0 -- because `length` is Int-valued.

-- Problem 20
-- Remove the k'th element from a list
removeAt :: Int -> [a] -> (a,[a])
removeAt k xs = let (l,r) = split xs (k-1) in (head r,l++tail r)
-- Technically this is unsafe if k is larger than the length of the list, but
-- that is an easy fix.


src/Q21t28.hs
module Q21t28 where

import Data.List (permutations)
import System.Random
import Q11t20 (removeAt)
import qualified Data.Bifunctor

tests :: Int -> Bool
tests 21 = insertAt 'X' "abcd" 2 == "aXbcd"
tests 22 = range 4 9 == [4,5,6,7,8,9]
tests 23 = True -- We cannot test random functions
tests 24 = True
tests 25 = True -- ditto
tests 26 = length (combinations 3 "abcdef") == choose' 6 3 where
    choose' _ 0 = 1
    choose' 0 _ = 0
    choose' m n = choose' (m-1) (n-1) * m `div` n
tests 27 = length (myGroup [2,3,4] "abcdefghi") == 1260 &&
            length (myGroup [2,2,5] "abcdefghi") == 756
tests 28 = isAscending (lsort theList) -- Part b would be a pain to test.
    where
        theList = ["abc","de","fgh","de","ijkl","mn","o"] 
        isAscending :: [[a]] -> Bool
        isAscending (x1:xs@(x2:xs'))
            | length x1 <= length x2 = isAscending xs
            | otherwise = False
        isAscending _ = True -- [] and [x] are ascending.
tests n
    | n <= 28 = True
    | otherwise = False

-- Problem 21
-- Insert an element at a given position into a list.
insertAt :: a -> [a] -> Int -> [a]
insertAt y xs n = let (l,r) = splitAt (n-1) xs in l ++ y:r

-- Problem 22
-- Create a list containing all integers within a given range
range :: Int -> Int -> [Int]
range n m
    | n <= m = take (m-n+1) [n..] -- add 1 because we want m to be in the list
    | otherwise = reverse $ take (n-m+1) [m..] -- go backwards if n > m

-- Problem 23
-- Extract a given number of randomly selected elements from a list

rndSelect :: (RandomGen g) => [a] -> Int -> g -> [a]
rndSelect [] _ _ = []
rndSelect xs 1 rng = let i = fst $ uniformR (0,length xs - 1) rng
    in [xs !! i]
rndSelect xs n rng
    | n > 1 = let (rng1,rng2) = split rng
        in rndSelect xs 1 rng1 ++ rndSelect xs (n-1) rng2
    | otherwise = []

rndSelectM :: [a] -> Int -> IO [a]
rndSelectM xs n = newStdGen >>= return . rndSelect xs n

rndSelectNoReplacement :: (RandomGen g) => [a] -> Int -> g -> [a]
rndSelectNoReplacement [] _ _ = []
rndSelectNoReplacement xs n rng
    | n == 1 = [snd $ sample rng xs]
    | n > 1 =
        let (rng1,rng2) = split rng in
            let (i0,x0) = sample rng1 xs
            in x0 : rndSelectNoReplacement (snd (removeAt (i0+1) xs)) (n-1) rng2
            -- We use i0+1 because removeAt 1 xs removes the first item, but
            -- i0 indexes from zero.
    | otherwise = []
    where sample rng' xs =
            let i0 = fst $ uniformR (0,length xs - 1) rng'
            in (i0,xs !! i0)

rndSelectNoReplacementM :: [a] -> Int -> IO [a]
rndSelectNoReplacementM xs n = newStdGen >>= return . rndSelectNoReplacement xs n

-- Problem 24
-- Draw n distinct random numbers from the set 1..m
diffSelect :: Int -> Int -> IO [Int]
diffSelect n m = rndSelectNoReplacementM [1..m] n

-- Problem 25
-- Generate a random permutation of the elements of a list
rndPermu :: [a] -> IO [a]
rndPermu xs = rndSelectNoReplacementM xs (length xs)

-- Problem 26
-- Generate the combinations of k distinct objects chosen from a list.
combinations :: Int -> [a] -> [[a]]
combinations 1 xs = map (: []) xs -- Send [x1,x2,..] to [[x1],[x2],..]
combinations _ [] = []
combinations n xs@(x:xs') =
    map (x : ) (combinations (n-1) xs') ++ combinations n xs'
    -- Adjoin x to all n-1 combinations which don't involve it, then take all n 
    -- combinations of things which don't involve x

-- Problem 27
-- Write a function which takes a list [a] and a list of integers [Int] and
-- outputs a list of disjoint subsets of [a] whose sizes are defined by [Int].
-- (these are the multinomial coefficients)
myGroup :: [Int] -> [a] -> [[[a]]]
myGroup [] _ = [[]]
myGroup (n:ns) xs = let parts = map f (comp' n xs) in concat parts
    where
        f = \(chunk,rest) -> map (chunk : ) (myGroup ns rest)
        -- comp' creates a new combinations function which outputs a tuple
        -- ([a],[a]) where the first component is the set of chosen objects,
        -- and the second component is the remainder of the list
        comp' :: Int -> [a] -> [([a],[a])]
        comp' _ [] = []
        comp' 1 (x:xs) = ([x],xs) :
            map (Data.Bifunctor.second (x :) ) (comp' 1 xs)
        comp' n xs@(x:xs') =
            map (Data.Bifunctor.first (x :) ) (comp' (n-1) xs') ++
                map (Data.Bifunctor.second (x :)) (comp' n xs')
    -- parts is the set of groups p of partitions s_i, indexed by splittings
    -- (chunk_i,rest_i) such that chunk_i is an element of s_i, and the union
    -- of all other s in s_i equals rest_i.  We concat to union all parititions
    -- together.

-- Problem 28a
-- Given a list of lists [[a]], sort it according to length, in ascending order.

{- | Takes a function `le` where le x y means x <= y, and sorts a list
     in ascending order 
-}
mySort :: (a -> a -> Bool) -> [a] -> [a]
mySort le (x1:xs)
    | x1_is_smallest = x1 : mySort le xs
    | otherwise = let (ys,zs) = splitAt j xs in mySort le (ys ++ x1:zs)
    where
        smallestIndex :: [Bool] -> Int
        smallestIndex [] = 0
        smallestIndex (x:xs) -- Find first False
            | not x = 1
            | otherwise = 1 + smallestIndex xs
        bools = map (x1 `le`) xs
        x1_is_smallest = and bools
        j = smallestIndex bools
mySort _ xs = xs -- Empty sets and singletons are already sorted.

lsort :: [[a]] -> [[a]]
lsort = mySort le
    where le x1 x2 = length x1 <= length x2

-- Problem 28b
-- Given a list of lists [[a]], sort it so that the least frequent lengths
-- appear first.
lfsort :: [[a]] -> [[a]]
lfsort xs = mySort lessFrequent xs
    where
        ls = map length xs
        frequency :: Int -> Int
        frequency l0 = foldr (\ l acc -> if l == l0 then 1 + acc else acc) 0 ls
        lessFrequent x1 x2 = frequency (length x1) <= frequency (length x2)


src/Q31t41.hs
module Q31t41 where

import Data.Tuple (swap)
import Q1t10 (encode)

{- | The unit tests for Problems 31 to 41 -}
tests :: Int -> Bool
tests 31 = isPrime 7
tests 32 = [myGCD 36 63, myGCD 3 6, myGCD 9 9] == [9,3,9]
tests 33 = coprime 35 64
tests 34 = totient 10 == 4
tests 35 = primeFactors 315 == [3,3,5,7]
tests 36 = primeFactorsMult 315 == [(3,2),(5,1),(7,1)]
tests 37 = totient' 10 == 4
tests 39 = primesR 10 20 == [11,13,17,19]
tests 40 = goldbach 28 == (5,23)
tests n
    | n <= 41 = True
    | otherwise = False

-- Problem 31
-- Determine whether a given integer number is prime
isPrime :: Int -> Bool
isPrime n
    | n == 2 = True
    | n > 2 = null [d | d <- [2..m], n `mod` d == 0]
    | otherwise = False
    where m = ceiling . sqrt $ fromIntegral n
    -- We need to separate the n=2 case because ceiling . sqrt $ fromIntegral 2
    -- is 2, so [d | d <- [2..m] ...] is nonempty.

-- The Sieve of Eratosthenes would speed a lot of these algorithms up.
-- We will create the list of primes naively, though.
primesR :: Int -> Int -> [Int]
primesR m n
    | m <= n = thePrimes
    | otherwise = reverse $ primesR n m
    where thePrimes = [p | p <- [m..n] , isPrime p]
primes :: [Int]
primes = [p | p <- [2..], isPrime p]

-- Problem 32
-- Find the gcd of two positive integers via Euclid's algorithm.
-- The Euclidean Algorithm calculates gcd m n as follows:
--      If m = q*n + r, then gcd m n = gcd n r.
-- Therefore, we repeatedly divide until r = 0, in which case case,
-- gcd m n = n.
myGCD :: Int -> Int -> Int
myGCD m n
    | m < n = myGCD n m -- Just to keep things easy to track
    | otherwise = if r == 0 then abs n else myGCD n r
    where r = mod m n

-- Problem 33
-- Determine whether two positive integers are coprime.
coprime :: Int -> Int -> Bool
coprime m n = 1 == gcd m n

-- Problem 34
-- Calculate Euler's totient gunction
totient :: Int -> Int
totient n = length [c | c <- [1..n], coprime n c]

-- Problem 35
-- Determine the prime factors of a positive integer. Output them in a list
-- in ascending order.
primeFactors :: Int -> [Int]
primeFactors n
    | n <= 1 = []
    | isPrime n = [n]
    | otherwise = p0 : primeFactors ( n `div` p0)
    where
        m = ceiling . sqrt $ fromIntegral n
        p0 = head [p | p <- primes, n `mod` p == 0]

-- Problem 36
-- Determine prime factors and give their multiplicity
primeFactorsMult :: Int -> [(Int,Int)]
primeFactorsMult = map swap . concatMap clean . encode . primeFactors
    -- We swap because the tuples from encode are (run_length,element)
    where -- we add a clean function to eliminate the Maybe monad from encode.
        clean :: (Int, Maybe Int) -> [(Int, Int)]
        clean (_,Nothing) = []
        clean (x,Just y) = [(x,y)]
        -- I wonder if there is a way to do this using functors

-- Problem 37
-- Use Problem 36 to calculate the totient
totientFormula :: [(Int,Int)] -> Int
totientFormula =
    foldr (\ (p,m) acc -> acc * ( (p - 1) * p ^ (m - 1) ) ) 1

totient' :: Int -> Int
totient' = totientFormula . primeFactorsMult
-- Alternatively one could use list comprehension:
-- totient' n = 
--     product [ (p - 1) * p ^ (m - 1) | (p,m) <- primeFactorsMult n]

-- Problem 38
-- Compare totient and totient'. Which one is faster?

-- Problem 39
-- List the prime numbers from m to n
-- See primesR above

-- Problem 40
-- Given an even integer, find two primes which sum to it
goldbach :: Int -> (Int,Int)
goldbach n
    | even n && n > 2 = head [s | s <- filteredPrimes, uncurry (+) s == n]
    | otherwise = (0,0) -- For odd numbers we just return something
    where
        filteredPrimes = [(p,n - p) | p <- primes, isPrime (n-p)]

-- Problem 41
-- Given a range of integers m and n, print the goldbach decompositions
-- for each even integer between m and n given from Problem 40.
goldbachList :: Int -> Int -> IO ()
goldbachList = error "Not Implemented"
